src/Table.cs

```diff
@@ -0,0 +1,72 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+using SpacetimeDB.BSATN;
+
+namespace SpacetimeDB
+{
+    public interface IDatabaseRow : IStructuralReadWrite { }
+
+    public abstract class RemoteBase<DbConnection>
+    {
+        protected readonly DbConnection conn;
+
+        protected RemoteBase(DbConnection conn)
+        {
+            this.conn = conn;
+        }
+    }
+
+    public interface IRemoteTableHandle
+    {
+        void SetCache(ClientCache.ITableCache cache);
+
+        object? GetPrimaryKey(IDatabaseRow row);
+
+        void InternalInvokeValueInserted(IDatabaseRow row);
+        void InternalInvokeValueDeleted(IDatabaseRow row);
+        void InvokeInsert(IEventContext context, IDatabaseRow row);
+        void InvokeDelete(IEventContext context, IDatabaseRow row);
+        void InvokeBeforeDelete(IEventContext context, IDatabaseRow row);
+        void InvokeUpdate(IEventContext context, IDatabaseRow oldRow, IDatabaseRow newRow);
+    }
+
+    public abstract class RemoteTableHandle<EventContext, Row> : IRemoteTableHandle
+        where EventContext : class, IEventContext
+        where Row : IDatabaseRow, new()
+    {
+        public void SetCache(ClientCache.ITableCache cache) => Cache = (ClientCache.TableCache<Row>)cache;
```

Comment:
We can leave this as a future work, but IIUC there is no reason for TableCache and RemoteTableHandle to be separate classes?

Previously all tables were static, which is why we needed a separate class for TableCache, but now that we have handles, we can just put cached items right into them.

Reply:
Ah that's a good point, yeah we can leave merging them for later.